File: app/src/main/java/com/github/ilikeyourhat/kudoku/solving/deduction/solver/DeductionSolver.java
package com.github.ilikeyourhat.kudoku.solving.deduction.solver;

import com.github.ilikeyourhat.kudoku.model.Sudoku;
import com.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid;
import com.github.ilikeyourhat.kudoku.solving.SudokuSolver;
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.DeductionAlgorithm;
import com.github.ilikeyourhat.kudoku.model.Sudoku;
import com.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid;
import com.github.ilikeyourhat.kudoku.solving.SudokuSolver;
import org.jetbrains.annotations.NotNull;
import com.github.ilikeyourhat.kudoku.model.SudokuType;

import java.util.List;

public abstract class DeductionSolver implements SudokuSolver {

    protected abstract List<DeductionAlgorithm.Factory> provideAlgorithms(SudokuType type);

    @NotNull
    @Override
    public Sudoku solve(@NotNull Sudoku sudoku) {
        sudoku = sudoku.copy();
        SudokuHintGrid sudokuHintGrid = new SudokuHintGrid(sudoku);
        return solve(sudoku, sudokuHintGrid);
    }

    private Sudoku solve(Sudoku sudoku, SudokuHintGrid sudokuHintGrid) {
        List<DeductionAlgorithm.Factory> algorithmFactories = provideAlgorithms(sudoku.getType());

        boolean gridHasChanged;
        do {
            gridHasChanged = false;
            for (DeductionAlgorithm.Factory factory : algorithmFactories) {
                if (factory.instance(sudoku.getRegions(), sudokuHintGrid).solve()){
                    gridHasChanged = true;
                    break;
                }
            }
        } while (gridHasChanged && !sudoku.isCompleted());
        return sudoku;
    }

}

File: app/src/main/kotlin/com/github/ilikeyourhat/kudoku/solving/deduction/solver/DeductionSolver.kt
package com.github.ilikeyourhat.kudoku.solving.deduction.solver

import com.github.ilikeyourhat.kudoku.model.Sudoku
import com.github.ilikeyourhat.kudoku.model.SudokuType
import com.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid
import com.github.ilikeyourhat.kudoku.solving.SudokuSolver
import com.github.ilikeyourhat.kudoku.solving.deduction.algorithm.DeductionAlgorithm

abstract class DeductionSolver : SudokuSolver {

    protected abstract fun provideAlgorithms(type: SudokuType): List<DeductionAlgorithm.Factory>

    override fun solve(sudoku: Sudoku): Sudoku {
        val copy = sudoku.copy()
        val sudokuHintGrid = SudokuHintGrid.create(copy)
        return solve(copy, sudokuHintGrid)
    }

    private fun solve(sudoku: Sudoku, sudokuHintGrid: SudokuHintGrid): Sudoku {
        val algorithmFactories = provideAlgorithms(sudoku.type)
        var gridHasChanged: Boolean
        do {
            gridHasChanged = false
            for (factory in algorithmFactories) {
                if (factory.instance(sudoku.regions, sudokuHintGrid).solve()) {
                    gridHasChanged = true
                    break
                }
            }
        } while (gridHasChanged && !sudoku.isCompleted())
        return sudoku
    }
}